source/utils.py

```diff
@@ -0,0 +1,110 @@
+        fh.write(rendered)
+
+    for plugin in plugins.values():
+        plugin_cli_name = plugin.name.replace('_', '-')
+        plugin_dir = os.path.join(rst_dir, plugin_cli_name)
+        os.mkdir(plugin_dir)
+
+        index_path = os.path.join(plugin_dir, 'index.rst')
+        with open(index_path, 'w') as fh:
+            template = env.get_template('plugin.rst')
+            rendered = template.render(title=plugin_cli_name, plugin=plugin)
+            fh.write(rendered)
+
+        if plugin.citations:
+            index_bib = os.path.join(plugin_dir, 'citations.bib')
+            write_bibtex(plugin.citations, index_bib)
+
+        for action in plugin.actions.values():
+            import_path, action_api_name = action.get_import_path().rsplit(
+                '.', 1)
+            action_cli_name = action.id.replace('_', '-')
+            action_path = os.path.join(plugin_dir, '%s.rst' % action_cli_name)
+
+            bib_id = ':'.join([plugin_cli_name, action_cli_name])
+            if action.citations:
+                action_bib = os.path.join(plugin_dir,
+                                          '%s.bib' % bib_id)
+                write_bibtex(action.citations, action_bib)
+
+            with open(action_path, 'w') as fh:
+                title = '%s: %s' % (action_cli_name, action.name)
+                directive_indent = ' ' * 3
+
+                logger.info("Generating help text for `%s %s`..." %
+                            (plugin_cli_name, action_cli_name))
+                command = ['qiime', plugin_cli_name, action_cli_name, '--help']
+                proc = subprocess.run(command, check=True,
+                                      stdout=subprocess.PIPE,
+                                      stderr=subprocess.STDOUT)
+                cli_help = textwrap.indent(proc.stdout.decode('utf-8'),
+                                           directive_indent).strip()
+                api_help = textwrap.indent(action.__call__.__doc__,
+                                           directive_indent).strip()
+
+                template = env.get_template('action.rst')
+                rendered = template.render(title=title, cli_help=cli_help,
+                                           api_help=api_help, bib_id=bib_id,
+                                           has_citations=bool(
+                                            action.citations),
+                                           import_path=import_path,
+                                           action_api_name=action_api_name)
+                fh.write(rendered)
+
+
+def cleanup_rst():
+    if os.path.exists(plugin_directory_rst_dir):
+        shutil.rmtree(plugin_directory_rst_dir)
+
+
+def write_bibtex(records, path):
+    citations = qiime2.sdk.Citations()
+    for idx, record in enumerate(records):
+        citations['key%d' % idx] = record
+    citations.save(path)
+
+
+def generate_plugin_dir():
+    generate_rst()
+    cleanup_rst()
```

Comment:
The `cleanup_rst()` function call probably needs to be deferred to some point later in the build process. In the `sphinx-ext-qiime2` implementation of this it is executed on the [`build-finished` event](https://www.sphinx-doc.org/en/master/extdev/appapi.html#event-build-finished). As implemented here the cleanup happens too soon, and the build errors:

```bash
$ make preview                                                                                                  build-plugin-dir faa5f1c ✗
rm -rf build/*
sphinx-build -b dirhtml -D command_block_no_exec=1 -D debug_page= -d build/doctrees -W source build/preview
Running Sphinx v3.3.1
Generating QIIME 2 plugin directory... (this may take a while)
QIIME is caching your current deployment for improved performance. This may take a few moments and should only happen once per deployment.
Generating help text for `vsearch cluster-features-de-novo`...
Generating help text for `vsearch cluster-features-closed-reference`...
Generating help text for `vsearch dereplicate-sequences`...
Generating help text for `vsearch join-pairs`...
Generating help text for `vsearch uchime-ref`...
Generating help text for `vsearch uchime-denovo`...
Generating help text for `vsearch fastq-stats`...
Generating help text for `vsearch cluster-features-open-reference`...
Generating help text for `taxa collapse`...
Generating help text for `taxa filter-table`...
Generating help text for `taxa filter-seqs`...
Generating help text for `taxa barplot`...
Generating help text for `sample-classifier regress-samples-ncv`...
Generating help text for `sample-classifier classify-samples-ncv`...
Generating help text for `sample-classifier fit-classifier`...
Generating help text for `sample-classifier fit-regressor`...
Generating help text for `sample-classifier predict-classification`...
Generating help text for `sample-classifier predict-regression`...
Generating help text for `sample-classifier split-table`...
Generating help text for `sample-classifier scatterplot`...
Generating help text for `sample-classifier confusion-matrix`...
Generating help text for `sample-classifier summarize`...
Generating help text for `sample-classifier classify-samples`...
Generating help text for `sample-classifier classify-samples-from-dist`...
Generating help text for `sample-classifier regress-samples`...
Generating help text for `sample-classifier metatable`...
Generating help text for `sample-classifier heatmap`...
Generating help text for `quality-filter q-score`...
Generating help text for `quality-control exclude-seqs`...
Generating help text for `quality-control filter-reads`...
Generating help text for `quality-control bowtie2-build`...
Generating help text for `quality-control evaluate-composition`...
Generating help text for `quality-control evaluate-seqs`...
Generating help text for `quality-control evaluate-taxonomy`...
Generating help text for `phylogeny midpoint-root`...
Generating help text for `phylogeny fasttree`...
Generating help text for `phylogeny raxml`...
Generating help text for `phylogeny raxml-rapid-bootstrap`...
Generating help text for `phylogeny iqtree`...
Generating help text for `phylogeny iqtree-ultrafast-bootstrap`...
Generating help text for `phylogeny filter-table`...
Generating help text for `phylogeny robinson-foulds`...
Generating help text for `phylogeny align-to-tree-mafft-fasttree`...
Generating help text for `phylogeny align-to-tree-mafft-iqtree`...
Generating help text for `phylogeny align-to-tree-mafft-raxml`...
Generating help text for `metadata distance-matrix`...
Generating help text for `metadata tabulate`...
Generating help text for `longitudinal nmit`...
Generating help text for `longitudinal first-differences`...
Generating help text for `longitudinal first-distances`...
Generating help text for `longitudinal pairwise-differences`...
Generating help text for `longitudinal pairwise-distances`...
Generating help text for `longitudinal linear-mixed-effects`...
Generating help text for `longitudinal anova`...
Generating help text for `longitudinal volatility`...
Generating help text for `longitudinal plot-feature-volatility`...
Generating help text for `longitudinal feature-volatility`...
Generating help text for `longitudinal maturity-index`...
Generating help text for `gneiss ilr-hierarchical`...
Generating help text for `gneiss ilr-phylogenetic`...
Generating help text for `gneiss correlation-clustering`...
Generating help text for `gneiss gradient-clustering`...
Generating help text for `gneiss assign-ids`...
Generating help text for `gneiss dendrogram-heatmap`...
Generating help text for `fragment-insertion sepp`...
Generating help text for `fragment-insertion classify-otus-experimental`...
Generating help text for `fragment-insertion filter-features`...
Generating help text for `feature-table rarefy`...
Generating help text for `feature-table subsample`...
Generating help text for `feature-table presence-absence`...
Generating help text for `feature-table relative-frequency`...
Generating help text for `feature-table transpose`...
Generating help text for `feature-table group`...
Generating help text for `feature-table merge`...
Generating help text for `feature-table merge-seqs`...
Generating help text for `feature-table merge-taxa`...
Generating help text for `feature-table rename-ids`...
Generating help text for `feature-table filter-samples`...
Generating help text for `feature-table filter-features-conditionally`...
Generating help text for `feature-table filter-features`...
Generating help text for `feature-table filter-seqs`...
Generating help text for `feature-table summarize`...
Generating help text for `feature-table tabulate-seqs`...
Generating help text for `feature-table core-features`...
Generating help text for `feature-table heatmap`...
Generating help text for `feature-classifier fit-classifier-sklearn`...
Generating help text for `feature-classifier classify-sklearn`...
Generating help text for `feature-classifier fit-classifier-naive-bayes`...
Generating help text for `feature-classifier extract-reads`...
Generating help text for `feature-classifier classify-consensus-blast`...
Generating help text for `feature-classifier classify-consensus-vsearch`...
Generating help text for `feature-classifier classify-hybrid-vsearch-sklearn`...
Generating help text for `emperor plot`...
Generating help text for `emperor procrustes-plot`...
Generating help text for `emperor biplot`...
Generating help text for `diversity pcoa`...
Generating help text for `diversity pcoa-biplot`...
Generating help text for `diversity procrustes-analysis`...
Generating help text for `diversity filter-distance-matrix`...
Generating help text for `diversity alpha-group-significance`...
Generating help text for `diversity bioenv`...
Generating help text for `diversity beta-group-significance`...
Generating help text for `diversity mantel`...
Generating help text for `diversity alpha-correlation`...
Generating help text for `diversity alpha-rarefaction`...
Generating help text for `diversity beta-rarefaction`...
Generating help text for `diversity adonis`...
Generating help text for `diversity beta-phylogenetic`...
Generating help text for `diversity beta`...
Generating help text for `diversity alpha-phylogenetic`...
Generating help text for `diversity alpha`...
Generating help text for `diversity core-metrics-phylogenetic`...
Generating help text for `diversity core-metrics`...
Generating help text for `diversity beta-correlation`...
Generating help text for `diversity-lib faith-pd`...
Generating help text for `diversity-lib observed-features`...
Generating help text for `diversity-lib pielou-evenness`...
Generating help text for `diversity-lib shannon-entropy`...
Generating help text for `diversity-lib bray-curtis`...
Generating help text for `diversity-lib jaccard`...
Generating help text for `diversity-lib unweighted-unifrac`...
Generating help text for `diversity-lib weighted-unifrac`...
Generating help text for `diversity-lib alpha-passthrough`...
Generating help text for `diversity-lib beta-passthrough`...
Generating help text for `diversity-lib beta-phylogenetic-passthrough`...
Generating help text for `diversity-lib beta-phylogenetic-meta-passthrough`...
Generating help text for `demux emp-single`...
Generating help text for `demux emp-paired`...
Generating help text for `demux subsample-single`...
Generating help text for `demux subsample-paired`...
Generating help text for `demux filter-samples`...
Generating help text for `demux summarize`...
Generating help text for `deblur denoise-16S`...
Generating help text for `deblur denoise-other`...
Generating help text for `deblur visualize-stats`...
Generating help text for `dada2 denoise-single`...
Generating help text for `dada2 denoise-paired`...
Generating help text for `dada2 denoise-pyro`...
Generating help text for `cutadapt trim-single`...
Generating help text for `cutadapt trim-paired`...
Generating help text for `cutadapt demux-single`...
Generating help text for `cutadapt demux-paired`...
Generating help text for `composition add-pseudocount`...
Generating help text for `composition ancom`...
Generating help text for `alignment mafft`...
Generating help text for `alignment mafft-add`...
Generating help text for `alignment mask`...
/Users/matthew/.conda/envs/2020.11-staging/lib/python3.6/site-packages/q2doc/external_links.py:11: RemovedInSphinx40Warning: The app.add_javascript() is deprecated. Please use app.add_js_file() instead.
  app.add_javascript('external-links.js')
making output directory... done
building [mo]: targets for 0 po files that are out of date
building [dirhtml]: targets for 44 source files that are out of date
updating environment: [new config] 44 added, 0 changed, 0 removed
-reading sources... [ 79%] tutorials/moving-pictures                                                                                        reading sources... [100%] tutorials/utilities

Warning, treated as error:
/Users/matthew/projects/qiime2/docs/source/plugins/index.rst:6:toctree contains reference to nonexisting document 'plugins/available/index'
make: *** [Makefile:48: preview] Error 2
```

I'm not too sure if there is a way to hook into the event system _outside_ of an extension. Perhaps you can modify the `Makefile` to clean up these files, instead?

Reply:
I decoupled the generation/clean-up functions and hooked the clean-up function to the Sphinx app/build-finished event again, so the behavior should be functionally equivalent.  See the travis build [here](https://travis-ci.com/github/andrewsanchez/docs/builds/209442101).